router/core/src/main/scala/io/buoyant/router/RoutingFactory.scala

```diff
@@ -3,22 +3,26 @@ package io.buoyant.router
 import com.twitter.finagle.{param => _, _}
 import com.twitter.finagle.buoyant.{Dst, DstBindingFactory}
 import com.twitter.finagle.tracing.Trace
-import com.twitter.util.{Future, Time}
+import com.twitter.util.{Throw, Return, Future, Time}
 import scala.util.control.NoStackTrace
 
 object RoutingFactory {
   val role = Stack.Role("RoutingFactory")
   val description = "Performs per-request name binding"
 
-  case class UnknownDst[Req](request: Req, cause: Throwable)
-    extends Exception(s"Unknown destination: $request / ${cause.getMessage}", cause)
+  case class UnknownDst[Req](request: Req, reason: String)
+    extends Exception(s"Unknown destination: $request / $reason")
     with NoStackTrace
 
+  sealed trait RequestIdentification[Req]
+  class IdentifiedRequest[Req](val dst: Dst, val request: Req) extends RequestIdentification[Req]
+  class UnidentifiedRequest[Req](val reason: String, val request: Req) extends RequestIdentification[Req]
```

Comment:
You're right, this is confusing because request is in `IdentifiedRequest` and `UnidentifiedRequest` for different reasons.  In `IdentifiedRequest`, the request is part of the output of identification.  In `UnidentifiedRequest` it's there for informational purposes.  I think we can remove it since the original request should be available in any place where you're getting an `UnidentifiedRequest`.